tests/channel_int/channel_int.c

```diff
@@ -0,0 +1,513 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+#include <stdint.h>
+
+#include "macro_utils/macro_utils.h"
+#include "testrunnerswitcher.h"
+
+#include "c_pal/interlocked.h"
+#include "c_pal/interlocked_hl.h"
+#include "c_pal/thandle.h"
+#include "c_pal/threadpool.h"
+#include "c_pal/threadapi.h"
+#include "c_pal/sync.h"
+
+#include "c_util/rc_ptr.h"
+#include "c_util/channel.h"
+
+TEST_DEFINE_ENUM_TYPE(CHANNEL_CALLBACK_RESULT, CHANNEL_CALLBACK_RESULT_VALUES);
+TEST_DEFINE_ENUM_TYPE(CHANNEL_RESULT, CHANNEL_RESULT_VALUES);
+TEST_DEFINE_ENUM_TYPE(THREADAPI_RESULT, THREADAPI_RESULT_VALUES);
+TEST_DEFINE_ENUM_TYPE(INTERLOCKED_HL_RESULT, INTERLOCKED_HL_RESULT_VALUES);
+
+#define CHANNEL_ORDER_TEST_COUNT 100000
+
+static struct
+{
+    THANDLE(THREADPOOL) g_threadpool;
+}g;
+
+static EXECUTION_ENGINE_HANDLE g_execution_engine = NULL;
+
+static volatile_atomic int32_t g_push_callback_count;
+
+static void* test_data = (void*)0x0001;
+static int32_t pull_cancelled = 0x0002;
+static int32_t push_cancelled = 0x0003;
+static int32_t pull_success = 0x0004;
+static int32_t push_success = 0x0005;
+static int32_t pull_abandoned = 0x0006;
+static int32_t push_abandoned = 0x0007;
+static void* test_data2 = (void*)0x0008;
+
+
+static void test_on_pull_callback_cancelled(void* context, CHANNEL_CALLBACK_RESULT result, THANDLE(RC_PTR) data)
```

Comment:
>test_on_pull_callback_cancelled [](http://example.com/codeflow?start=12&length=31)

these "test" callbacks would be so much more easier to read if they'd have a int32_t* original = context, just so that the reader would understand what the original context used to be at the time when the function was called.